### PR TITLE
WIP: new sample reference shaders 

### PR DESCRIPTION
**Pull request by jiradeto**
Related issue: #545

Implemented 310 es sample shaders:

**1. trigonometric_strip** 

trigonometric_strip
- invokes a number of trigonometric built-in functions
- have a ternary operator

**2. selection_sort_struct**

selection_sort_struct
- have a struct data type
- have an array as the attribute of struct
- perform the selection sort

**3. prefix_sum_checkers**

prefix_sum_checkers
- derived from the prefix_sum shader with a bit of modification
- use uniform many times
- have while loop insides for loop

### CODE REVIEW

File: shaders/src/main/glsl/samples/310es/prefix_sum_checkers.frag [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+precision highp float;
+
+layout(location = 0) out vec4 _GLF_color;
+
+uniform vec2 injectionSwitch;
+
+uniform vec2 resolution;
+
+vec2 pattern( in vec2 x )
+{
```
Comment (afd) [inline shaders/src/main/glsl/samples/310es/prefix_sum_checkers.frag]: Weird indentation here.

File: shaders/src/main/glsl/samples/310es/prefix_sum_checkers.frag [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	} else if (int(gl_FragCoord.y) < 180) {
+	  	col = .5 + cos( c.y  + vec3( resolution.x, A[44]/resolution.x + 50., 22.0) );	
+  	} else if (int(gl_FragCoord.y) < 200) {
+      	col = .5 + cos( c.y  + vec3( resolution.x, A[49]/resolution.x + 50., 22.0) );		  
+  	} else {
+		discard;
+  	}
+	
+	_GLF_color = vec4(col, 1.0);
+}
```
Comment (afd) [inline shaders/src/main/glsl/samples/310es/prefix_sum_checkers.frag]: Probably best to have new line at end of file.

File: shaders/src/main/glsl/samples/310es/selection_sort_struct.frag [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  vec3 col =  tan(pow(uv.xxx, uv.yyy)  + 
+  vec3(
+    obj.odd_numbers[int(floor(gl_FragCoord.x / 1000.))],
+    obj.even_numbers[int(floor(gl_FragCoord.y / 1000.))],
+    sinh(uv.x)
+  ));
+  
+  _GLF_color.rgb = col;
+  _GLF_color.a = 1.0;
+}
```
Comment (afd) [inline shaders/src/main/glsl/samples/310es/selection_sort_struct.frag]: New line

File: shaders/src/main/glsl/samples/310es/prefix_sum_checkers.frag [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+		}
+	}
+
+	vec2 c = pattern( (15.0 + tan(0.2)) * uv);
+	vec3 col;
+ 	if (int(gl_FragCoord.y) < 20) {
+		col = .5 + cos(c.y + vec3(esolution.x, A[4]/resolution.x + 50., 22.0));
+	} else if (int(gl_FragCoord.y) < 40) {
+	  	col = .5 + cos(c.y + vec3(esolution.x, A[9]/resolution.x + 50., 22.0));	
+  	} else if (int(gl_FragCoord.y) < 60) {
```
Comment (paulthomson) [inline shaders/src/main/glsl/samples/310es/prefix_sum_checkers.frag]: Two typos above: esolution -> resolution

File: shaders/src/main/glsl/samples/310es/selection_sort_struct.frag [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  Obj obj;
+  
+  // Initialize first 10 odd numbers to the array 
+  int odd_index = 0;
+  float odd_number = 1.;
+  while(odd_index <= 10){
+    obj.odd_numbers[odd_index] = odd_number;
+    odd_number += 2.;
+    odd_index++;
+  }
```
Comment (paulthomson) [inline shaders/src/main/glsl/samples/310es/selection_sort_struct.frag]: I think there is an out of bound access here: `obj.odd_numbers[odd_index]` where `odd_index` is 10.

File: shaders/src/main/glsl/samples/310es/prefix_sum_checkers.frag (hunk 4 above)
Comment (paulthomson) [reply shaders/src/main/glsl/samples/310es/prefix_sum_checkers.frag]: Run shaders: `glslangValidator X.frag`

File: shaders/src/main/glsl/samples/310es/selection_sort_struct.frag (hunk 5 above)
Comment (jiradeto) [reply shaders/src/main/glsl/samples/310es/selection_sort_struct.frag]: Thank you! Fixed

File: shaders/src/main/glsl/samples/310es/prefix_sum_checkers.frag (hunk 4 above)
Comment (jiradeto) [reply shaders/src/main/glsl/samples/310es/prefix_sum_checkers.frag]: Fixed

File: shaders/src/main/glsl/samples/310es/prefix_sum_checkers.frag (hunk 2 above)
Comment (jiradeto) [reply shaders/src/main/glsl/samples/310es/prefix_sum_checkers.frag]: Updated

File: shaders/src/main/glsl/samples/310es/prefix_sum_checkers.frag [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+uniform vec2 injectionSwitch;
+
+uniform vec2 resolution;
+
+vec2 pattern(in vec2 x){
+	vec2 n = floor(x);
+	vec3 m = vec3(1.);
+	
+    for (int j = -1; j <= int(injectionSwitch.y); j++) {
+		for (int i = -1; i <= int(injectionSwitch.y); i++) {
```
Comment (paulthomson) [inline shaders/src/main/glsl/samples/310es/prefix_sum_checkers.frag]: Indentation above.

File: shaders/src/main/glsl/samples/310es/selection_sort_struct.frag [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  // Initialize first 10 odd numbers to the array.
+  int odd_index = 0;
+  float odd_number = 1.0;
+  while (odd_index <= 9) {
+    obj.odd_numbers[odd_index] = odd_number;
+    odd_number += 2.0;
+    odd_index++;
+  }
+  // Similarly, initizlie even numbers and iterate backward.
+  int even_index = 9;
```
Comment (paulthomson) [inline shaders/src/main/glsl/samples/310es/selection_sort_struct.frag]: Spelling above.

Comment (jiradeto) [reply shaders/src/main/glsl/samples/310es/selection_sort_struct.frag]: Fixed!

File: shaders/src/main/glsl/samples/310es/prefix_sum_checkers.frag (hunk 6 above)
Comment (jiradeto) [reply shaders/src/main/glsl/samples/310es/prefix_sum_checkers.frag]: The indentation is now using spaces instead of tabs.

**Pull request merged**